# mCREATE

Example code for generating analyses and plots for the multiplexed CREATE paper by Kumar et al. (2019). The code is divided into three parts: alignment, variant analysis, and clustering.

## Alignment

### Prerequisites

Alignment uses Python 3 and Jupyter notebooks, and depends on the following Python packages:
```
jupyter
pepars
protfarm
```

Pepars and Protfarm are available at their respective GitHub repositories:
- Pepars: [https://github.com/GradinaruLab/pepars]("https://github.com/GradinaruLab/pepars")
- Protfarm: [https://github.com/GradinaruLab/protfarm]("https://github.com/GradinaruLab/protfarm")

### Running

To run an example alignment, see the example script ```alignment/align_and_export.ipynb```. This scripts initializes a Protfarm data workspace, downloads example FASTQ files, aligns them against a template, and calculates and exports the count and enrichment data. For more details on Protfarm workspaces and enrichment, see the [Protfarm]("https://github.com/GradinaruLab/protfarm") repository.

## Variant Analysis

Variant analysis contains a set of scripts that operate on count and enrichment Excel sheets, as generated by aligning and exporting data via Protfarm.

### Prerequisites

Variant analysis requires the following Python packages:
```
jupyter
pepars
```

## Clustering

Clustering analyses use sequence and count data as output from Protfarm, or from manual analysis, to analyze 

### Prerequisites

Clustering analysis requires Matlab 2018 or later.

### Running

####hammDist.m
This script is a Matlab function that generates a .txt list of paired sequences with their reverse hamming distance, 
i.e. number of positions with identical AA. Output may be visualized in Cytoscape software.